File: apps/frontend/src/lib/constants/physics.ts
// Physics constants - MUST match backend values exactly
export const PHYSICS = {
    GRAVITY: 800.0, // pixels/s²
    TERMINAL_VELOCITY: 600.0, // pixels/s
    JUMP_STRENGTH: -400.0, // pixels/s (negative = upward)
    GROUND_LEVEL: 500.0, // Y coordinate
    DEFAULT_SPEED: 200.0, // pixels/s
    FIXED_STEP: 1.0 / 60.0, // seconds — MUST match backend FixedDeltaTime exactly
    RECONCILIATION_SMOOTHING_DECAY: 0.85, // per-frame multiplier for visual offset decay
    DASH_SPEED: 600.0, // pixels/s
    DASH_DURATION: 0.12, // seconds
    DASH_COOLDOWN_TIME: 0.5, // seconds
    DOWN_DASH_SPEED: 800.0, // pixels/s
    DOWN_DASH_DURATION: 0.25, // seconds
    BLOCK_SPEED_FACTOR: 0.3, // movement speed multiplier while blocking
    ATTACK_DURATION: 0.2, // seconds — active attack swing duration
    ATTACK_COOLDOWN_TIME: 0.4, // seconds — minimum time between attacks
    ATTACK_WIDTH: 40.0, // pixels — horizontal reach of attack hitbox
    ATTACK_HEIGHT: 30.0, // pixels — vertical height of attack hitbox
    BODY_WIDTH: 40.0, // pixels — player rectangle width (must match backend BodyWidth)
    BODY_HEIGHT: 40.0, // pixels — player rectangle height (must match backend BodyHeight)
    PARRY_WINDOW: 0.1, // seconds — perfect deflect window after pressing block
    MAX_POSTURE: 100.0, // posture break threshold
    BLOCK_POSTURE_DAMAGE: 15.0, // posture build-up on defender from blocking a hit
    HIT_POSTURE_DAMAGE: 30.0, // posture build-up on defender from a raw hit
    DEFLECT_POSTURE_DAMAGE: 25.0, // posture damage reflected to attacker on perfect deflect
    POSTURE_RECOVERY_RATE: 20.0, // posture decay per second when idle
    MAX_HEALTH: 100.0, // starting health pool
    STAGGER_DURATION: 1.2, // stun duration in seconds when posture breaks
    HIT_HP_DAMAGE: 20.0, // health damage from a raw hit
    CRITICAL_HP_DAMAGE: 50.0, // health damage dealt to a staggered player
} as const;

File: apps/backend/src/constants/physics.go
package constants

const (
	// Gravity is the downward acceleration applied per second (pixels/s²)
	Gravity = 800.0

	// TerminalVelocity is the maximum falling speed (pixels/s)
	TerminalVelocity = 600.0

	// JumpStrength is the upward velocity applied when jumping (pixels/s)
	JumpStrength = -400.0

	// GroundLevel is the Y coordinate of the ground/floor
	GroundLevel = 500.0

	// DefaultSpeed is the horizontal movement speed (pixels/s)
	DefaultSpeed = 200.0

	// FixedDeltaTime is the exact physics step duration in seconds.
	// Both server and client MUST use this identical value for all
	// physics simulation to prevent floating-point drift.
	FixedDeltaTime = 1.0 / 60.0

	// DashSpeed is the horizontal speed during a dash (pixels/s)
	DashSpeed = 600.0

	// DashDuration is how long the dash lasts in seconds
	DashDuration = 0.12

	// DashCooldownTime is the minimum time between dashes in seconds
	DashCooldownTime = 0.5

	// DownDashSpeed is the downward plunge speed (pixels/s)
	DownDashSpeed = 800.0

	// DownDashDuration is the maximum duration of a down-dash before timeout
	DownDashDuration = 0.25

	// BlockSpeedFactor is the movement speed multiplier while blocking (0.3 = 30% speed)
	BlockSpeedFactor = 0.3

	// AttackDuration is the active attack swing duration in seconds
	AttackDuration = 0.2

	// AttackCooldownTime is the minimum time between initiating attacks in seconds
	AttackCooldownTime = 0.4

	// AttackWidth is the horizontal reach of the attack hitbox in pixels
	AttackWidth = 40.0

	// AttackHeight is the vertical height of the attack hitbox in pixels
	AttackHeight = 30.0

	// BodyWidth is the player rectangle width in pixels (must match frontend BODY_W)
	BodyWidth = 40.0

	// BodyHeight is the player rectangle height in pixels (must match frontend BODY_H)
	BodyHeight = 40.0

	// ParryWindow is the time in seconds after pressing block during which a
	// successful hit counts as a perfect deflect instead of a normal block.
	ParryWindow = 0.1

	// MaxPosture is the posture threshold at which a player becomes staggered.
	MaxPosture = 100.0

	// BlockPostureDamage is the posture build-up on the defender from blocking a hit.
	BlockPostureDamage = 15.0

	// HitPostureDamage is the posture build-up on the defender from taking a raw hit.
	HitPostureDamage = 30.0

	// DeflectPostureDamage is the posture damage reflected back to the attacker on a perfect deflect.
	DeflectPostureDamage = 25.0

	// PostureRecoveryRate is the posture decay rate per second when idle (not blocking, not attacking).
	PostureRecoveryRate = 20.0
)
